parachain/primitives/ethereum/src/ethashproof.rs

```diff
@@ -0,0 +1,353 @@
+        }
+    }
+
+    fn dag_merkle_root(&self, epoch: u64) -> Option<H128> {
+        DAGS_MERKLE_ROOTS.get((epoch - DAGS_START_EPOCH) as usize)
+            .map(|x| H128::from(x))
+    }
+
+    // Adapted fro https://github.com/near/rainbow-bridge/blob/3fcdfbc6c0011f0e1507956a81c820616fb963b4/contracts/near/eth-client/src/lib.rs#L363
+    pub fn hashimoto_merkle(
+        &self,
+        header_hash: H256,
+        nonce: H64,
+        header_number: u64,
+        nodes: &[DoubleNodeWithMerkleProof],
+    ) -> Result<(H256, H256), &'static str> {
+        // Check that we have the expected number of nodes with proofs
+        const MIXHASHES: usize = MIX_BYTES / HASH_BYTES;
+        if nodes.len() != MIXHASHES * ACCESSES / 2 {
+            return Err("Incorrect number of nodes");
+        }
+    
+        let epoch = header_number / EPOCH_LENGTH;
+        // Reuse single Merkle root across all the proofs
+        let merkle_root = self.dag_merkle_root(epoch).ok_or("Invalid epoch")?;
+        let full_size = ethash::get_full_size(epoch as usize);
+
+        // Boxed index since ethash::hashimoto gets Fn, but not FnMut
+        let index = RefCell::new(0);
+        // Flag for whether the proof is valid
+        let success = RefCell::new(true);
+
+        let results = ethash::hashimoto_with_hasher(
+            header_hash,
+            nonce,
+            full_size,
+            |offset| {
+                let idx = *index.borrow_mut();
+                *index.borrow_mut() += 1;
+
+                // Each two nodes are packed into single 128 bytes with Merkle proof
+                let node = &nodes[idx / 2];
+                if idx % 2 == 0 {
+                    // Divide by 2 to adjust offset for 64-byte words instead of 128-byte
+                    if let Ok(computed_root) = node.apply_merkle_proof((offset / 2) as u64) {
+                        if merkle_root != computed_root {
+                            success.replace(false);
+                        }
+                    } else {
+                        success.replace(false);
+                    }
+                };
+
+                // Reverse each 32 bytes for ETHASH compatibility
+                let mut data = node.dag_nodes[idx % 2].0;
+                data[..32].reverse();
+                data[32..].reverse();
+                data.into()
+            },
+            keccak_256,
+            keccak_512,
+        );
+
+        match success.into_inner() {
+            true => Ok(results),
+            false => Err("Invalid merkle proof"),
+        }
+    }
+
+    pub fn hashimoto_light(
+        &mut self,
+        header_hash: H256,
+        nonce: H64,
+        header_number: u64,
+    ) -> (H256, H256) { 
+        let epoch = header_number / EPOCH_LENGTH;
+        let cache = match self.dags_cache {
+            Some(ref mut c) => c.get(epoch, header_number),
+            None => panic!("EthashProver wasn't configured with hashimoto light cache"),
```

Comment:
One more panic here. I know the `hashimoto_light` function isn't being used anywhere yet, but lets fix it now so we don't accidently cause problems in the future.

Reply:
@vgeddes i'm totally guessing, but this panic looks like it would only occur if the cache is not set, which should only happen if the chain is not initialized correctly or if the cache expires after a few years, in which case the bridge would actually be broken and we likely do want a panic causing the parachain to halt?

@Rizziepit can you confirm?

Reply:
@musnit correct-ish. It simply means the struct wasn't initialized using the below:
```
let prover = EthashProver::with_hashimoto_light(10);
```

I coded it like this because I don't want the `hashimoto_light` function to work by default. An engineer needs to first decide what size cache they can tolerate. There's also the issue that the `EthashProver` needs to be a thread-safe singleton for the cache to be re-used. I didn't delve into how to implement that because I expect Substrate will complicate the matter.

Suffice to say, that panic indicates engineering error and cannot be caused by user input. The node is unusable. I'm not sure `Result` is appropriate. Thoughts, @vgeddes?

Reply:
Sounds good. I understand now :+1: